File: go/backtracking/77_combine.go
package backtracking

import "slices"

// combine 给定两个整数 n 和 k，返回范围 [1, n] 中所有可能的 k 个数的组合。
func combine(n int, k int) [][]int {

	// 保存所有结果的集合
	res := make([][]int, 0)
	// 满足条件的路径
	path := make([]int, 0, k)

	// 回溯算法， startIdx 代表从n 中开始取的数字
	var backtrack func(int)
	backtrack = func(startIdx int) {

		if len(path) == k {
			p := slices.Clone(path)
			res = append(res, p)
			return
		}

		// 优化剪枝
		// i 选择之后，剩余的元素个数应该至少k个，才有遍历的意义，比如n=4,k=4 那至少循环开始之后有4个元素可选择
		// 所以 i <= n - (k - len(path)) + 1
		// for i := startIdx; i < n+1; i++ {
		for i := startIdx; i <= n-(k-len(path))+1; i++ {
			path = append(path, i)

			backtrack(i + 1)
			// 回溯
			path = path[:len(path)-1]
		}
	}

	backtrack(1)

	return res
}

File: rust/src/backtracking/_77_combine.rs
/// combine 给定两个整数 n 和 k，返回范围 [1, n] 中所有可能的 k 个数的组合。
pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
    // 所有满足条件的集合
    let mut res = vec![];
    // 当前满足条件的路径
    let mut path = vec![];

    fn backtrack(start_idx: i32, n: i32, k: i32, path: &mut Vec<i32>, res: &mut Vec<Vec<i32>>) {
        if path.len() == k as usize {
            res.push(path.clone());
            return;
        }

        for i in start_idx..=(n - (k - path.len() as i32) + 1) {
            path.push(i);
            backtrack(i + 1, n, k, path, res);
            path.pop();
        }
    }

    backtrack(1, n, k, &mut path, &mut res);
    res
}
